From 8b0391ca61d0a606e9829b95cf67620c78e0cd61 Mon Sep 17 00:00:00 2001
From: chenyutao <chenyutao2@huawei.com>
Date: Thu, 5 Feb 2026 21:06:22 +0800
Subject: [PATCH] urma: Fix urma_perftest size check for send opc

1. change default send size order from 16 to 12;
2. fix size check for send opcode, at most 64k for RTP;
   4k for CTP.

Signed-off-by: chenyutao <chenyutao2@huawei.com>
(cherry picked from commit f3aa171532c8ea60ea73c065f98049b2f4675642)
---
 .../tools/urma_perftest/perftest_parameters.c | 34 +++++++++++++++++--
 1 file changed, 32 insertions(+), 2 deletions(-)

diff --git a/src/urma/tools/urma_perftest/perftest_parameters.c b/src/urma/tools/urma_perftest/perftest_parameters.c
index 9769ad9..f04c42e 100644
--- a/src/urma/tools/urma_perftest/perftest_parameters.c
+++ b/src/urma/tools/urma_perftest/perftest_parameters.c
@@ -26,6 +26,11 @@
 #define PERFTEST_JFC_MUL_THRESHOLD (4)
 #define PERFTEST_DEFAULT_DURATION (5)
 
+#define PERFTEST_RTP_MAX_SEND_SIZE (65536)
+#define PERFTEST_CTP_MAX_SEND_SIZE (4096)
+#define PERFTEST_RTP_MAX_ORDER     (16)
+#define PERFTEST_CTP_MAX_ORDER     (12)
+
 typedef struct perftest_cmd {
     char *cmd;
     perftest_cmd_type_t type;
@@ -83,7 +88,8 @@ static void usage(const char *argv0)
 {
     command_usage(argv0);
     (void)printf("Options:\n");
-    (void)printf("  -a, --all[order]            Run sizes from 2 till 2^23 (default 2^16), order: exponent of 2.\n");
+    (void)printf("  -a, --all[order]            Run sizes from 2 till 2^23,\n"
+                 "                              default 2^12 for send, 2^16 for others, order: exponent of 2.\n");
     (void)printf("  -A, --atomic_type <type>    Specify atomic type, {cas|faa}.\n");
     (void)printf("  -b, --simplex_mode          Run with simplex mode(jfs/jfr), duplex jetty mode for reserved.\n");
     (void)printf("  -B, --bidirection           Measure bidirectional bandwidth (default unidirectional).\n");
@@ -283,7 +289,7 @@ static void init_cfg(perftest_config_t *cfg)
     cfg->jfc_inline = false;
     cfg->inf_period = PERFTEST_DEF_INF_PERIOD;
     cfg->inf_period_ms = 0;
-    cfg->order = PERFTEST_SIZE_ORDER;
+    cfg->order = (cfg->api_type == PERFTEST_SEND) ? PERFTEST_CTP_MAX_ORDER : PERFTEST_SIZE_ORDER;
     cfg->err_timeout = URMA_TYPICAL_ERR_TIMEOUT;
     cfg->lock_free = false;
     cfg->priority = URMA_MAX_PRIORITY;
@@ -1262,6 +1268,30 @@ int check_local_cfg(perftest_config_t *cfg)
         (void)fprintf(stderr, "UM transport mode is not recommended for ctp.\n");
     }
 
+    if (cfg->api_type == PERFTEST_SEND) {
+        if (cfg->single_path && cfg->size > PERFTEST_RTP_MAX_SEND_SIZE && !cfg->all) {
+            (void)fprintf(stderr, "Invalid size: %u with single path for send opcode, max size: %u.\n",
+                cfg->size, PERFTEST_RTP_MAX_SEND_SIZE);
+            exit(1);
+        }
+        if (!cfg->single_path && cfg->size > PERFTEST_CTP_MAX_SEND_SIZE && !cfg->all) {
+            (void)fprintf(stderr, "Invalid size: %u with multi path for send opcode, max size: %u.\n",
+                cfg->size, PERFTEST_CTP_MAX_SEND_SIZE);
+            exit(1);
+        }
+
+        if (cfg->single_path && cfg->order > PERFTEST_RTP_MAX_ORDER) {
+            (void)fprintf(stderr, "Invalid order: %u with single path for send opcode, max order: %u.\n",
+                cfg->order, PERFTEST_RTP_MAX_ORDER);
+            exit(1);
+        }
+        if (!cfg->single_path && cfg->order > PERFTEST_CTP_MAX_ORDER) {
+            (void)fprintf(stderr, "Invalid order: %u with multi path for send opcode, max order: %u.\n",
+                cfg->order, PERFTEST_CTP_MAX_ORDER);
+            exit(1);
+        }
+    }
+
     return 0;
 }
 
-- 
2.28.0.windows.1

